content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown

```diff
@@ -0,0 +1,150 @@
+    geom_bar(width = 1) +
+    coord_polar(theta = "y") +
+    xlab(NULL) +
+    ylab(NULL)
+}
+piechart_basic(mpg, aes(factor(1), fill = class))
+```
+
+By quoting and unquoting the arguments, the user can leave the arguments bare when calling the function.
+
+```{r piechart-qq}
+piechart <- function(data, var, ...) {
+  var <- enquo(var)
+  piechart_basic(data, aes(!!factor(1), fill = !!var))
+}
+piechart(mpg, class)
+```
+
+The basic pattern is to enquote and unquote named arguments.
+
+```{r scatter-by}
+scatter_by <- function(data, x, y) {
+  x <- enquo(x)
+  y <- enquo(y)
+
+  ggplot(data) + geom_point(aes(!!x, !!y))
+}
+scatter_by(mtcars, disp, drat)
+```
+
+### Facetting with `vars()`
+
+To support quasiquotation in facetting we’ve added a new helper function: [`vars()`](https://ggplot2.tidyverse.org/reference/vars.html), short for variables. Instead of `facet_grid(x + y ~ a + b)` you can now write `facet_grid(vars(x, y), vars(a, b))`. The formula interface won’t go away; but the new `vars()` interface supports tidy evaluation, so can be easily programmed with.
+
+`vars()` is used to supply variables or expressions, evaluated in the context of the dataset to form facetting groups.
+
+```{r facet-vars}
+p <- ggplot(mpg, aes(displ, cty)) + geom_point()
+
+p + facet_grid(rows = vars(drv))
+```
+
+Using quosures ensures that the variable comes from the context of the dataframe. Since the ellipsis is the first and only argument of `vars()`, you can use unquote splicing with the `!!!` operator to pass in a list of arguments.
+
+```{r vars-env}
+year <- 2018
+
+d <- mpg %>%
+  filter(manufacturer %in% c("chevrolet", "dodge", "ford", "toyota")) %>%
+  ggplot() +
+    geom_point(aes(displ, cty))
+
+args <- list(quo(year), quo(manufacturer))
+
+d + facet_grid(vars(!!!args))
+```
+
+Inside of `vars()` you can easily supply names, which will add titles to the facets.
+
+```{r labelled-grid}
+p + facet_grid(vars(Cylinder = cyl), labeller = label_both)
+```
+
+### Using `vars()` for wrapper functions
+
+`vars()` makes it easier to pass variables from wrapper functions.
+
+```{r wrap-by}
+p <- ggplot(mtcars, aes(wt, disp)) + geom_point()
+
+wrap_by <- function(...) {
+  facet_wrap(vars(...), labeller = label_both)
+}
+
+p + wrap_by(vs, am)
+```
+
+In our `wrap_by()` function above, we used tidy dots ([`...`](https://adv-r.hadley.nz/quasiquotation.html#dot-dot-dot-...)), which represent an arbitrary number of additional arguments. For a function to use *named arguments*, we'll need to quote the named argument with [`enquo()`](http://rlang.r-lib.org/reference/quotation.html). To create a default name, we'll use [`quo_name()`](http://rlang.r-lib.org/reference/quo_label.html), which transforms a quosure into a simple string. Then we unquote and evaluate our arguments in their proper contexts using the [`!!`](http://rlang.r-lib.org/reference/quasiquotation.html) (read: bang bang) operator, and the `:=` operator to unquote the name.
```

Comment:
For a function to... -> "Alternatively, we could allow the user to provide a single named argument..."